### ISSUE TITLE
Option to run a playbook fully in parallel on many hosts

### ISSUE THREAD

**Issue by hkratz** [ansible/ansible#2551]
As far as I can see right now Ansible waits until a task completes on all hosts before it moves on to the next task. It would be great if we had an option to run a playbook completely in parallel on all hosts. This will reduce the time for running a playbook.

**Comment by bcoca** [comment]
This should not be the case, only if you set serial=1.

Brian Coca

**Comment by hkratz** [comment]
@bcoca The individual tasks are executed in parallel on all hosts. But if one host finishes a task early the next task is not started for this host. Instead Ansible waits until all hosts have finished the task, then the next is started (in parallel on all hosts again) and so on.

**Comment by mpdehaan** [comment]
this is true, and a feature.

**Comment by btiernay** [comment]
Seems to be a limiting feature:

http://stackoverflow.com/questions/21158689/ansible-deploy-on-multiple-hosts-in-the-same-time

Any chance it can be generalized a bit to support this common use case?

**Comment by runesl** [comment]
Often a node hangs for a while on a specific task, and all the other have to wait. Then for the next task another node may hang for a while, again requiring everybody to wait for a single node.

There should be an option to make the tasks of a playbook run in parallel across the nodes. 
I.e. force sequential task execution for a single node, but not across nodes. A LOT of use cases would benefit from this.

**Comment by bkeroackdsc** [comment]
+1. This would be very nice to have. This is an advantage salt has over Ansible, since in that system the tasks are compiled and ran on the end servers ("minions") autonomously.

**Comment by echo-devnull** [comment]
Well, theoretically you could use a playbook to do nothing else than to setup "ansible pull mode" and have those servers install themselves independently from the other tasks in you play. "In parallel" so to speak.

**Comment by tetherit** [comment]
+1 - I have hundreds of servers, some on very slow Edge connections - all the servers on fast connections have to wait many hours for the servers on slow ones :(

**Comment by abadger** [comment]
This is something that we're looking to put the foundation for into v2.
The actual ability to fully parallelize may or may not make it in depending
on how complete it looks at the time we release but if nothing else the
rewrite should make it possible to add it (whereas the current code base
would make this a monumental undertaking).

-Toshio

**Comment by btiernay** [comment]
Perhaps we can reopen the ticket then?

**Comment by rahulmehrotra1** [comment]
Would love to see this feature; would help reduce deployment time. Specially when there are no dependencies between included: playbooks

**Comment by lielran** [comment]
+1 
when you are running hundreds of servers deployment and all of them is waiting from single node to finish this feature become very useful

**Comment by bcoca** [comment]
this is now in devel, choose the free strategy for your play.

**Comment by lielran** [comment]
great news! 
@bcoca - can you point me to docs/how to use it ?

**Comment by bcoca** [comment]
We implemented the feature, still writing the docs ....

On Mon, Jun 8, 2015 at 2:02 PM, Liel Ran notifications@github.com wrote:

> great news!
> @bcoca https://github.com/bcoca - can you point me to docs/how to use
> it ?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ansible/ansible/issues/2551#issuecomment-110093349.

## 

Brian Coca

**Comment by tetherit** [comment]
Any news on where to find the docs? :)

**Comment by aleemb** [comment]
+1 would be swell. A pointer to the PR or commit would be helpful in lieu of the docs until they become available

**Comment by bcoca** [comment]
http://docs.ansible.com/ansible/playbooks_strategies.html

**Comment by MarkRosen** [comment]
Saw an example of "strategy"  today:
```
- hosts: eng
  stategy: free
  tasks: ...
```
Full disclosure , just saw the above syntax in a slide during a demo from RedHat.... I haven't personally validated it.

**Comment by anastazya** [comment]
> Saw an example of "strategy" today:
> 
> ```
> - hosts: eng
>   stategy: free
>   tasks: ...
> ```
> Full disclosure , just saw the above syntax in a slide during a demo from RedHat.... I haven't personally validated it.

Not working : "ERROR! 'stategy' is not a valid attribute for a Play"

**Comment by roelandjansen** [comment]
I assume it's missing the 'r' in the example. E.g. "strategy"
